fleetspeak/src/server/https/client_certificate.go

```diff
@@ -1,32 +1,87 @@
 package https
 
 import (
+	"crypto/sha256"
 	"crypto/x509"
+	"encoding/base64"
+	"encoding/hex"
 	"encoding/pem"
 	"errors"
 	"fmt"
 	"net/http"
 	"net/url"
+	"strings"
 
 	cpb "github.com/google/fleetspeak/fleetspeak/src/server/components/proto/fleetspeak_components"
+	log "github.com/golang/glog"
 )
 
 // GetClientCert returns the client certificate from either the request header or TLS connection state.
-func GetClientCert(req *http.Request, hn string, frontendMode cpb.FrontendMode) (*x509.Certificate, error) {
+func GetClientCert(req *http.Request, hn string, frontendMode cpb.FrontendMode, chcksumHeader string) (*x509.Certificate, error) {
 	switch frontendMode {
 	case cpb.FrontendMode_MTLS:
-		if hn == "" {
+		if hn == "" && chcksumHeader == "" {
 			return getCertFromTLS(req)
 		}
 	case cpb.FrontendMode_HEADER_TLS:
-		if hn != "" {
-			return getCertFromHeader(hn, req.Header)
+		if hn != "" && chcksumHeader == "" {
+			return getCertFromHeader(hn, req.Header, "", false)
+		}
+	case cpb.FrontendMode_HEADER_TLS_CHECKSUM:
+		if (hn != "" && chcksumHeader != "") {
+			return getCertFromHeader(hn, req.Header, chcksumHeader, true)
 		}
 	}
-	return nil, fmt.Errorf("received invalid frontend mode combination: frontendMode=%s, clientCertHeader=%s", frontendMode, hn)
+	return nil, fmt.Errorf("received invalid frontend mode combination: frontendMode=%s, clientCertHeader=%s, clientCertHeaderChecksum=%s",
+				frontendMode, hn, chcksumHeader)
+}
+
+func calcClientCertSha256(clientCert string) (string) {
+	// Decode the PEM string
+	block, _ := pem.Decode([]byte(clientCert))
```

Comment:
`func Decode(data []byte) (p *Block, rest []byte)`, the second return value is a byte slice that would contain other blocks. 

Maybe it's worth verifying that the rest is empty, because we don't expect there to be another block after the certficate.

Reply:
Added len(rest) to verify that rest is empty.